`make` fails when building as root

Confirmed on both Ubuntu 20.04 and Ubuntu 24.04:

```bash
# the error seems to only occur if we are root
[shwestrick@ubuntu]$ sudo bash
[root@ubuntu]$ make
...
make compiler CHECK_FIXPOINT=false                     # mlton0 -> mlton1
make[1]: Entering directory '/home/shwestrick/proj/mlton'
make -C "/home/shwestrick/proj/mlton/mlton"
make[2]: Entering directory '/home/shwestrick/proj/mlton/mlton'
(									\
	echo '$(SML_LIB)/basis/unsafe.mlb';				\
	echo '$(SML_LIB)/basis/sml-nj.mlb';				\
	echo '$(SML_LIB)/basis/mlton.mlb';				\
	echo '$(SML_LIB)/basis/basis.mlb';				\
	mlbdeps() { if [ -e "$1" ]; then mlton -stop f "$1"; else echo "$1"; fi; } && mlbdeps mlton.mlb |		\
		grep -v 'mlb$' |					\
		grep 'mlyacc';					\
	mlbdeps() { if [ -e "$1" ]; then mlton -stop f "$1"; else echo "$1"; fi; } && mlbdeps ../lib/stubs/mlton-stubs/sources.mlb |		\
		grep -v 'mlb$' |					\
		grep 'mlton-stubs';					\
	mlbdeps() { if [ -e "$1" ]; then mlton -stop f "$1"; else echo "$1"; fi; } && mlbdeps mlton.mlb |		\
		grep -v 'mlb$' |					\
		grep -v 'sml/basis' |				\
		grep -v 'targets' |					\
		grep -v 'mlyacc';					\
) > mlton-stubs.mlb
rm -rf front-end/ml.lex.sml
cp -fpR front-end/ml.lex.sml.boot front-end/ml.lex.sml
chmod -w front-end/ml.lex.sml
chmod: front-end/ml.lex.sml: new permissions are r--rw-r--, not r--r--r--
make[2]: *** [Makefile:117: front-end/ml.lex.sml] Error 1
make[2]: Leaving directory '/home/shwestrick/proj/mlton/mlton'
make[1]: *** [Makefile:72: compiler] Error 2
make[1]: Leaving directory '/home/shwestrick/proj/mlton'
make: *** [Makefile:19: all] Error 2
```

The error is a failing call to `chmod -w ...`. It appears this is due to a strange edge-case behavior of `chmod`. Some helpful threads [here](https://askubuntu.com/questions/975331/why-isnt-chmod-doing-what-i-want-it-to) and [here](https://askubuntu.com/questions/718777/strange-behaviour-of-sudo-chmod-rwx-does-not-remove-group-write-permissions/718782#718782).

To be honest, I don't fully understand the problem, in part because I can't figure out why it's failing now when it hasn't in the past. I encountered this when attempting to build a Docker image that built MLton from source, and I've built this Docker image before (within the past year) with no issues.

# Fix

The solution appears to be to use `chmod a-w` instead of `chmod -w`. I've already confirmed (on Ubuntu 24.04) that this causes the `chmod` error to go away.

I can throw together a quick PR now.

--- a/mlton/Makefile
+++ b/mlton/Makefile
@@ -109,12 +109,12 @@ front-end/%.lex.sml: front-end/%.lex
 	$(SED) -e 's/in Vector.fromList(List.map g/in Vector.fromList(Pervasive.List.map g/' $<.sml.in > $<.sml
 	$(RM) $<.sml.in
 	$(CP) $<.sml $<.sml.boot
-	$(CHMOD) -w $<.*
+	$(CHMOD) a-w $<.*
 else
 front-end/%.lex.sml: front-end/%.lex.sml.boot
 	$(RM) $@
 	$(CP) $< $@
-	$(CHMOD) -w $@
+	$(CHMOD) a-w $@
 endif
 
 ifeq (true,$(call HAVE_CMD,$(RUN_MLYACC)))
@@ -131,16 +131,16 @@ front-end/%.grm.sig front-end/%.grm.sml: front-end/%.grm
 	$(SED) -e 's/in Array.fromList(List.map actionRowLookUp actionRowNumbers)/in Array.fromList(Pervasive.List.map actionRowLookUp actionRowNumbers)/' $<.sml.in > $<.sml
 	$(RM) $<.sml.in
 	$(CP) $<.sml $<.sml.boot
-	$(CHMOD) -w $<.*
+	$(CHMOD) a-w $<.*
 else
 front-end/%.grm.sig: front-end/%.grm.sig.boot
 	$(RM) $@
 	$(CP) $< $@
-	$(CHMOD) -w $@
+	$(CHMOD) a-w $@
 front-end/%.grm.sml: front-end/%.grm.sml.boot
 	$(RM) $@
 	$(CP) $< $@
-	$(CHMOD) -w $@
+	$(CHMOD) a-w $@
 endif
 
 mlton-stubs.mlb: $(shell $(MLBDEPS) ../lib/stubs/mlton-stubs/sources.mlb | $(GREP) 'mlb$$' | $(GREP) -v '$$(SML_LIB)') $(shell $(MLBDEPS) mlton.mlb | $(GREP) 'mlb$$' | $(GREP) -v '$$(SML_LIB)')

